transactor/src/main/scala/io/mediachain/datastore/DynamoDatastore.scala

```diff
@@ -0,0 +1,158 @@
+import scala.collection.mutable.{Buffer, ArrayBuffer}
+import io.mediachain.multihash.MultiHash
+import io.mediachain.types.Datastore.DatastoreException
+
+// TODO error handling
+//      queued eventual writes in the background with disk backing of 
+//       in-progress writes
+//      use batch writes for chunked puts
+class DynamoDatastore(table: String, creds: BasicAWSCredentials)
+  extends BinaryDatastore with AutoCloseable {
+  import scala.collection.JavaConversions._
+  
+  val chunkSize = 1024 * 384 // 384 KB; DynamoDB has 400KB limit
+  val chunkTable = table + "Chunks"
+  
+  val db = new AmazonDynamoDBClient(creds)
+
+  override def put(key: MultiHash, value: Array[Byte]) {
+    if (value.length < chunkSize) {
+      putSimple(key, value)
+    } else {
+      putChunked(key, value)
+    }
+  }
+
+  private def putSimple(key: MultiHash, value: Array[Byte]) {
+    val keyAttr = new AttributeValue
+    keyAttr.setS(key.base58)
+    
+    val dataAttr = new AttributeValue
+    dataAttr.setB(bytes2Buffer(value))
+    
+    val item = Map(
+      "multihash" -> keyAttr,
+      "data"      -> dataAttr
+    )
+    
+    db.putItem(table, item)
+  }
+
+  private def putChunked(key: MultiHash, value: Array[Byte]) {
+    val key58 = key.base58
+    val chunks = chunkCount(value.length)
+    val chunkIds = (1 to chunks).map { n => key58 + "#" + n }
+    
+    // this should use the BatchWriteItemRequest API (PITA alert)
+    chunkIds.zipWithIndex.foreach {
+      case (chunkId, index) => {
+        val keyAttr = new AttributeValue
+        keyAttr.setS(chunkId)
+
+        val dataAttr = new AttributeValue
+        dataAttr.setB(bytes2BufferChunk(value, index))
+        
+        val item = Map(
+          "chunkId" -> keyAttr,
+          "data" -> dataAttr
+        )
+        
+        db.putItem(chunkTable, item)
+      }
+    }
+    
+    val keyAttr = new AttributeValue
+    keyAttr.setS(key58)
+    
+    val chunkAttr = new AttributeValue
+    chunkAttr.setSS(chunkIds)
+    
+    val item = Map(
+      "multihash" -> keyAttr,
+      "chunks" -> chunkAttr
+      )
+    
+    db.putItem(table, item)
+  }
+
+  private def chunkCount(len: Int): Int = {
+    val div = len / chunkSize
+    val rem = len % chunkSize
+    if (rem > 0) {
+      div + 1
+    } else {
+      div
+    }
+  }
+  
+  private def bytes2Buffer(bytes: Array[Byte]) = 
+    ByteBuffer.wrap(bytes)
+  
+  private def bytes2BufferChunk(bytes: Array[Byte], chunk: Int) = {
+    val offset = chunk * chunkSize
+    val len = Math.min(chunkSize, bytes.length - offset)
+    ByteBuffer.wrap(bytes.slice(offset, offset + len))
+  }
+  
+  override def close() {
+    db.shutdown()
+  }
+
+  override def get(key: MultiHash): Option[Array[Byte]] = {
+    val keyAttr = new AttributeValue
+    keyAttr.setS(key.base58)
+    
+    val item = Map(
+      "multihash" -> keyAttr
+    )
+    val res = db.getItem(table, item).getItem  // yes, really.
+    
+    Option(res).map { item =>
```

Comment:
The logic is that first you check if you have a "data" attribute, which stores the binary directly.
If this is null, then you check for a chunks attribute which is a list of chunk-ids to fetch from the Chunks table.
So the value that you get out of the chunks attribute does not substitute the value of the data attribute for the monadic approach to work.


Also, there should be no case of having a NullPointer exception the way it is written, and I really don't want to swallow this.